### PR TITLE
Adds the procedure to import manifest list

### PR DESCRIPTION
**Pull request by stevsmit**
Adds the procedure to import manifest list

Versions: 4.12+

Issue: https://issues.redhat.com/browse/OSDOCS-4279

Pending QE/SME

Preview: https://52173--docspreview.netlify.app/openshift-enterprise/latest/openshift_images/image-streams-manage.html#importing-manifest-list-through-imagestreamimport_image-streams-managing

### CODE REVIEW

File: modules/importing-manifest-list-through-imagestreamimport.adoc [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+[NOTE]
+====
+You can compare the output from the aforementioned commands against the content of `quay.io/<user_name>/<image_name>`. If correct, there should be one image for the manifest list, and one image for each sub-manifest. 
+====
+
+. (Optiona) Delete the images previously created by entering the following command:
++
+[source,terminal]
+----
+$ oc delete images $(oc get images | grep quay.io/<user_name>/<image_name> | awk '{ print $1 }')
```
Comment (dmage) [inline modules/importing-manifest-list-through-imagestreamimport.adoc:60]: Please don't mention `oc delete images` in docs. It creates an illusion that the image is deleted, but only `oc adm prune images` can correctly delete image objects

* without breaking existing image streams
* with the removal of unnecessary data from the image registry storage

If you delete an image object via `oc delete` you can have leftovers on the storage that will never be deleted.

**Comment by dmage** [comment]
That's the hard way to do it, the easier way would be:

1. Use `oc edit` on an existing image stream.
2. Add

    ```
    importPolicy:
      importMode: "ManifestList"
    ```

    to tags that should be imported as a manifest list.

3. Wait until it gets imported.

**Comment by dmage** [comment]
we have two import modes: Legacy and PreserveOriginal
https://github.com/openshift/api/blob/4226c2167e40c62b09e2db083d1c0051fafba7da/image/v1/types.go#L300-L310

So it should be:

```
importPolicy:
  importMode: PreserveOriginal
```

File: modules/importing-manifest-list-through-imagestreamimport.adoc [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+. Enter the following command to ensure that all sub-manifests were created:
++
+[source,terminal]
+----
+$ oc get is app
+----
++
+[NOTE]
+====
+You can compare the output from the aforementioned commands against the content of `quay.io/<user_name>/<image_name>`. If correct, there should be one image for the manifest list, and one image for each sub-manifest.
```
Comment (flavianmissi) [inline modules/importing-manifest-list-through-imagestreamimport.adoc:53]: I'm unsure about the usefulness of this note to the end-user... Shouldn't they be able to trust that the image stream was created correctly without the need to verify it? Perhaps this is better as a troubleshooting step? 🤔 
I mostly added it to the [step-by-step](https://github.com/flavianmissi/imagestreams/blob/main/manifest-list-import.md) to aid testing.

File: modules/importing-manifest-list-through-imagestreamimport.adoc [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+  images:
+  - from:
+      kind: DockerImage
+      name: quay.io/<user_name>/<image_name>
+    to:
+      name: latest
+    referencePolicy:
+      type: Source
+    importPolicy:
+      importMode: "ManifestList"
```
Comment (flavianmissi) [inline modules/importing-manifest-list-through-imagestreamimport.adoc:33]: As Oleg mentioned, `ManifestList` is now `PreserveOriginal`

File: modules/importing-manifest-list-through-imagestreamimport.adoc [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+kind: ImageStreamImport
+metadata:
+  name: app
+  namespace: <namespace>
+spec:
+  import: true
+  images:
+  - from:
+      kind: DockerImage
+      name: quay.io/<user_name>/<image_name>
```
Comment (dmage) [inline modules/importing-manifest-list-through-imagestreamimport.adoc:27]: what do we usually have in docs when we need a container image? `quay.io/<user_name>/<image_name>` looks like only quay is supported here.

File: modules/importing-manifest-list-through-imagestreamimport.adoc [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+----
+
+. Enter the following command to ensure that Image objects were created for the manifest list: 
++
+[source,terminal]
+----
+$ oc get image | grep quay.io/<user_name>/<image_name>
+----
+
+. Enter the following command to ensure that all sub-manifests were created:
```
Comment (dmage) [inline modules/importing-manifest-list-through-imagestreamimport.adoc:44]: the provided command doesn't tell you about sub-manifests, and most users don't care how it is imported; the thing they care about - the image reference from `from.name` will be resolved into an image reference with sha and it'll work on all architectures that are supported by the source image.

File: modules/importing-manifest-list-through-imagestreamimport.adoc [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+====
+
+. (Optiona) Delete the images previously created by entering the following command:
++
+[source,terminal]
+----
+$ oc delete images $(oc get images | grep quay.io/<user_name>/<image_name> | awk '{ print $1 }')
+----
+
+. (Optional) delete the ImageStream by entering the following command:
```
Comment (dmage) [inline modules/importing-manifest-list-through-imagestreamimport.adoc:63]: If the image stream existed before these steps and had some other images, this command will delete them too. Perhaps we don't need to tell how to undo previous steps.

File: modules/importing-manifest-list-through-imagestreamimport.adoc (hunk 1 above)
Comment (stevsmit) [reply modules/importing-manifest-list-through-imagestreamimport.adoc:60]: @dmage should I replace `oc delete images` with `oc adm prune images $(oc adm prune images | grep quay.io/<user_name>/<image_name> | awk '{ print $1 }')` ?

File: modules/importing-manifest-list-through-imagestreamimport.adoc (hunk 4 above)
Comment (stevsmit) [reply modules/importing-manifest-list-through-imagestreamimport.adoc:27]: @dmage sometimes we use 

name: inputimage:latest.

WDYT? we don't have a value that's broader than that, from what I can tell. This is my first time writing for ImageStreams, though

File: modules/importing-manifest-list-through-imagestreamimport.adoc (hunk 1 above)
Comment (dmage) [reply modules/importing-manifest-list-through-imagestreamimport.adoc:60]: The pruner doesn't have any parameters, it works with all objects. And the pruner can be used only by admins, while everybody can import an image. So

* delete image stream tags/image streams that you don't need
* wait for some time (by default - 1 hour)
* run the pruner `oc adm prune images`

The wait time is also tricky, it can be changed via `--keep-younger-than`, but if you set it to a low value (like 5 seconds), this may corrupt ongoing image pushes.

So overall - normal users cannot interact with `image` objects, and admins shouldn't interact with them directly unless they really know what they are doing.

File: modules/importing-manifest-list-through-imagestreamimport.adoc (hunk 5 above)
Comment (stevsmit) [reply modules/importing-manifest-list-through-imagestreamimport.adoc:44]: So I can remove 

. Enter the following command to ensure that all sub-manifests were created:
+
[source,terminal]
----
$ oc get is app
----

?

File: modules/importing-manifest-list-through-imagestreamimport.adoc (hunk 6 above)
Comment (dmage) [reply modules/importing-manifest-list-through-imagestreamimport.adoc:63]: Yes, let's keep it simple and remove optional cleanup steps.

We may want to explain what is outcome of creating ImageStreamImport, but today we don't have anything like this, even for normal imports. AFAIK we don't have any docs about ImageStreamImport.

**Review by flavianmissi — COMMENTED** [review]
@stevsmit we need to address an inconsistency before moving forward here: image stream imports and image streams are different things.

`oc edit imagestream <your_imagestream>` will open up the ImageStream object for editing, not an ImageStreamImport as the docs seem to suggest.

ImageStreamImports aren't objects (aka resource types) like ImageStreams are, they are [virtual resource types](https://kubernetes.io/docs/reference/using-api/api-concepts/#standard-api-terminology) (aka virtual objects).
ImageStreamImports cannot be edited, while ImageStreams can. They can't be retrieved either, so `oc get` does not work for them. They can only be created. You can see in the [API docs for ImageStreamImports](https://docs.openshift.com/container-platform/4.11/rest_api/image_apis/imagestreamimport-image-openshift-io-v1.html#api-endpoints) that only POST is listed. As far as I understand ImageStreamImports aren't even stored on etcd. This isn't very common and it's quite confusing if you ask me.

In short, instead of `oc edit imagestream <your_imagestream>` the user needs to run `oc create -f <your_imagestreamimport.yaml>`. The yaml you're using will work. The image stream doesn't need to exist for the `oc create` command to work (`.spec.import: true` will auto-create the imagestream when it doesn't already exist).

File: modules/importing-manifest-list-through-imagestreamimport.adoc [hunk 7]
```diff
@@ -0,0 +1,48 @@
+// Module included in the following assemblies:
+// * assembly/openshift_images/managing-image-streams.adoc
+
+:_content-type: PROCEDURE
+[id="importing-manifest-list-through-imagestreamimport_{context}"]
+= Importing a manifest list through ImageStreamImport
+
+The `ImageStreamImport` resource provides users a way to find and import image manifests from other container image registries into the server. Individual images or an entire image repository can be imported. Users can choose to see the results of the import prior to tagging the resulting images into the specified image stream.
```
[the remark this answers is not in the document]
Comment (stevsmit) [reply modules/importing-manifest-list-through-imagestreamimport.adoc:8]: I'm not even sure where "Users can choose to see the results of the import prior to tagging the resulting images into the specified image stream." came from. I just removed it. I think I got it from somewhere else in the docs.

**Review by flavianmissi — APPROVED** [review]

File: modules/importing-manifest-list-through-imagestreamimport.adoc [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+  images:
+  - from:
+      kind: DockerImage
+      name: <registry>/<user_name>/<image_name>
+    to:
+      name: latest
+    referencePolicy:
+      type: Source
+    importPolicy:
+      importMode: "PreserveOriginal"
```
Comment (xiuwang) [inline modules/importing-manifest-list-through-imagestreamimport.adoc:35]: Should we mention another value "Legacy" for importMode here to explain the importMode usage? @dmage @flavianmissi

Comment (flavianmissi) [reply modules/importing-manifest-list-through-imagestreamimport.adoc:35]: That wouldn't hurt. "Legacy" is the default value, if nothing else we could just say that...?

Comment (stevsmit) [reply modules/importing-manifest-list-through-imagestreamimport.adoc:35]: @flavianmissi @xiuwang PTAL at the table I added to the section below. Thanks.

**Review by flavianmissi — APPROVED** [review]

**Comment by xiuwang** [comment]
/lgtm
Thanks

File: modules/importmode-configuration-fields.adoc [hunk 9]
```diff
@@ -0,0 +1,23 @@
+// Module included in the following assemblies:
+// * assembly/openshift_images/managing-image-streams.adoc
```
Comment (adellape) [inline modules/importmode-configuration-fields.adoc:2]: Looks like the file path for the assembly would be:

```
// * openshift_images/image-streams-manage.adoc
```

File: modules/importing-manifest-list-through-imagestreamimport.adoc [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+
+Use the following procedure to import a manifest list through the `ImageStreamImport` object with the `importMode` value.  
+
+.Procedure
+
+. Create an `ImageStreamImport` YAML file and set the `importMode` parameter to `PreserveOriginal` on the tags that you will import as a manifest list:
++
+[source,yaml]
+----
+---
```
Comment (jeana-redhat) [inline modules/importing-manifest-list-through-imagestreamimport.adoc:19]: Are the lines with `---` (here, and L37) intentional?
